File: Python/8-Graphs(BFS-DFS)/Algorithms/7-Union-Find(DisjointSetUnion-DSU)/Union-Find(DisjointSetUnion-DSU).py
"""
Union-Find (Disjoint Set Union - DSU) Implementation in Python

This module provides implementations of the Union-Find data structure, also known as
Disjoint Set Union.  It's useful for problems involving connected components in graphs,
detecting cycles, and efficiently grouping elements into sets.

Five approaches are included:

1.  Basic Union-Find:  Simple implementation with path compression but no explicit rank.
2.  Union-by-Rank:  Improves efficiency by attaching smaller trees to larger ones.
3.  Path Compression: Optimizes find operation by flattening the tree structure.
4.  Union-by-Rank with Path Compression: Combines both optimizations for best performance.
5.  Iterative Path Compression: A non-recursive version of path compression

Each approach includes:
    -   Initialization:  Creating the data structure.
    -   Find:  Finding the representative (root) of the set an element belongs to.
    -   Union:  Merging two sets.
    -   isConnected: Checks if two elements are in the same set.

Example Usage:
    # Initialize Union-Find with 5 elements
    uf = UnionFind(5)  #  or use UnionFindRanked(5), etc.

    # Perform some union operations
    uf.union(0, 1)
    uf.union(2, 3)
    uf.union(1, 3)  # Merges sets {0, 1} and {2, 3}

    # Check if elements are connected
    print(uf.isConnected(0, 3))  # Output: True
    print(uf.isConnected(0, 2))  # Output: False

    # Find the representative of an element
    print(uf.find(0))      # Might output 0 or 1 or 2 or 3 (depending on path compression)
    print(uf.find(2))      # Might output 2 or 3
"""


class UnionFind:
    """
    Basic Union-Find implementation with path compression.
    """

    def __init__(self, n):
        """
        Initializes the Union-Find data structure.

        Args:
            n: The number of elements.
        """
        self.parent = list(range(n))  # Initialize each element as its own parent

    def find(self, x):
        """
        Finds the representative (root) of the set that an element belongs to,
        with path compression.

        Args:
            x: The element.

        Returns:
            The representative of the set.
        """
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])  # Path compression
        return self.parent[x]

    def union(self, x, y):
        """
        Merges the sets containing elements x and y.

        Args:
            x: The first element.
            y: The second element.
        """
        root_x = self.find(x)
        root_y = self.find(y)
        if root_x != root_y:
            self.parent[root_x] = root_y  # Merge by attaching one root to another

    def isConnected(self, x, y):
        """
        Checks if two elements are in the same set.

        Args:
            x: The first element.
            y: The second element.

        Returns:
            True if the elements are in the same set, False otherwise.
        """
        return self.find(x) == self.find(y)

    def count_sets(self):
        """
        Counts the number of distinct sets.

        Returns:
            The number of distinct sets.
        """
        roots = set()
        for i in range(len(self.parent)):
            roots.add(self.find(i))
        return len(roots)


class UnionFindRanked:
    """
    Union-Find implementation with union by rank and path compression.
    """

    def __init__(self, n):
        """
        Initializes the Union-Find data structure.

        Args:
            n: The number of elements.
        """
        self.parent = list(range(n))
        self.rank = [0] * n  # Initialize rank of each element to 0

    def find(self, x):
        """
        Finds the representative (root) of the set that an element belongs to,
        with path compression.

        Args:
            x: The element.

        Returns:
            The representative of the set.
        """
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])  # Path compression
        return self.parent[x]

    def union(self, x, y):
        """
        Merges the sets containing elements x and y, using union by rank.

        Args:
            x: The first element.
            y: The second element.
        """
        root_x = self.find(x)
        root_y = self.find(y)
        if root_x != root_y:
            if self.rank[root_x] < self.rank[root_y]:
                self.parent[root_x] = root_y  # Attach smaller rank tree to larger
            elif self.rank[root_x] > self.rank[root_y]:
                self.parent[root_y] = root_x
            else:
                self.parent[root_y] = root_x  # If ranks are equal, choose either
                self.rank[root_x] += 1  # Increment the rank of the new root

    def isConnected(self, x, y):
        """
        Checks if two elements are in the same set.

        Args:
            x: The first element.
            y: The second element.

        Returns:
            True if the elements are in the same set, False otherwise.
        """
        return self.find(x) == self.find(y)

    def count_sets(self):
        """
        Counts the number of distinct sets.

        Returns:
            The number of distinct sets.
        """
        roots = set()
        for i in range(len(self.parent)):
            roots.add(self.find(i))
        return len(roots)



class UnionFindPathCompression:
    """
    Union-Find implementation with only path compression (no rank).  Included for completeness,
    though UnionFindRanked is generally preferred.  This is the same as the first
    implementation (UnionFind).
    """

    def __init__(self, n):
        """
        Initializes the Union-Find data structure.
        Args:
            n: The number of elements.
        """
        self.parent = list(range(n))

    def find(self, x):
        """
        Finds the representative (root) of the set that an element belongs to,
        with path compression.
        Args:
            x: The element.
        Returns:
            The representative of the set.
        """
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, x, y):
        """
        Merges the sets containing elements x and y.
        Args:
            x: The first element.
            y: The second element.
        """
        root_x = self.find(x)
        root_y = self.find(y)
        if root_x != root_y:
            self.parent[root_x] = root_y

    def isConnected(self, x, y):
        """
        Checks if two elements are in the same set.
        Args:
            x: The first element.
            y: The second element.
        Returns:
            True if the elements are in the same set, False otherwise.
        """
        return self.find(x) == self.find(y)

    def count_sets(self):
        """
        Counts the number of distinct sets.
        Returns:
        The number of distinct sets
        """
        roots = set()
        for i in range(len(self.parent)):
            roots.add(self.find(i))
        return len(roots)



class UnionFindRankedPathCompression:
    """
    Union-Find implementation with both union by rank and path compression.
    This is generally the most efficient implementation.
    """

    def __init__(self, n):
        """
        Initializes the Union-Find data structure.

        Args:
            n: The number of elements.
        """
        self.parent = list(range(n))
        self.rank = [0] * n

    def find(self, x):
        """
        Finds the representative (root) of the set that an element belongs to,
        with path compression.

        Args:
            x: The element.

        Returns:
            The representative of the set.
        """
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, x, y):
        """
        Merges the sets containing elements x and y, using union by rank.

        Args:
            x: The first element.
            y: The second element.
        """
        root_x = self.find(x)
        root_y = self.find(y)
        if root_x != root_y:
            if self.rank[root_x] < self.rank[root_y]:
                self.parent[root_x] = root_y
            elif self.rank[root_x] > self.rank[root_y]:
                self.parent[root_y] = root_x
            else:
                self.parent[root_y] = root_x
                self.rank[root_x] += 1

    def isConnected(self, x, y):
        """
        Checks if two elements are in the same set.

        Args:
            x: The first element.
            y: The second element.

        Returns:
            True if the elements are in the same set, False otherwise.
        """
        return self.find(x) == self.find(y)

    def count_sets(self):
        """
        Counts the number of distinct sets.

        Returns:
            The number of distinct sets.
        """
        roots = set()
        for i in range(len(self.parent)):
            roots.add(self.find(i))
        return len(roots)


class UnionFindIterativePC:
    """
    Union-Find implementation with iterative path compression and union by rank.
    This version of path compression uses a while loop instead of recursion.
    """

    def __init__(self, n):
        """
        Initializes the Union-Find data structure.

        Args:
            n: The number of elements.
        """
        self.parent = list(range(n))
        self.rank = [0] * n

    def find(self, x):
        """
        Finds the representative (root) of the set that an element belongs to,
        using *iterative* path compression.

        Args:
            x: The element.

        Returns:
            The representative of the set.
        """
        path = []
        root = x
        while root != self.parent[root]:
            path.append(root)
            root = self.parent[root]

        # Iterative path compression:
        for node in path:
            self.parent[node] = root
        return root

    def union(self, x, y):
        """
        Merges the sets containing elements x and y, using union by rank.

        Args:
            x: The first element.
            y: The second element.
        """
        root_x = self.find(x)
        root_y = self.find(y)
        if root_x != root_y:
            if self.rank[root_x] < self.rank[root_y]:
                self.parent[root_x] = root_y
            elif self.rank[root_x] > self.rank[root_y]:
                self.parent[root_y] = root_x
            else:
                self.parent[root_y] = root_x
                self.rank[root_x] += 1

    def isConnected(self, x, y):
        """
        Checks if two elements are in the same set.

        Args:
            x: The first element.
            y: The second element.

        Returns:
            True if the elements are in the same set, False otherwise.
        """
        return self.find(x) == self.find(y)

    def count_sets(self):
        """
        Counts the number of distinct sets.
        Returns:
            The number of distinct sets.
        """
        roots = set()
        for i in range(len(self.parent)):
            roots.add(self.find(i))
        return len(roots)


def main():
    """
    Main function to demonstrate the usage of the Union-Find implementations.
    """
    # Example usage with UnionFindRankedPathCompression (the most efficient)
    num_elements = 6
    print(f"\n--- Union-Find with {num_elements} elements ---")
    uf = UnionFindRankedPathCompression(num_elements)

    print(f"Initial parent array: {uf.parent}")  # Print initial state
    print(f"Initial ranks: {uf.rank}")

    # Perform union operations
    uf.union(0, 1)
    print(f"After union(0, 1): parent = {uf.parent}, ranks = {uf.rank}")
    uf.union(2, 3)
    print(f"After union(2, 3): parent = {uf.parent}, ranks = {uf.rank}")
    uf.union(4, 5)
    print(f"After union(4, 5): parent = {uf.parent}, ranks = {uf.rank}")
    uf.union(1, 3)  # Merges {0, 1} and {2, 3}
    print(f"After union(1, 3): parent = {uf.parent}, ranks = {uf.rank}")
    uf.union(0, 4)
    print(f"After union(0, 4): parent = {uf.parent}, ranks = {uf.rank}")

    # Check connectivity
    print(f"Is 0 connected to 3? {uf.isConnected(0, 3)}")  # Expected: True
    print(f"Is 0 connected to 2? {uf.isConnected(0, 2)}")  # Expected: True
    print(f"Is 0 connected to 5? {uf.isConnected(0, 5)}")  # Expected: True
    print(f"Is 1 connected to 4? {uf.isConnected(1, 4)}")  # Expected: True
    print(f"Is 2 connected to 5? {uf.isConnected(2, 5)}")  # Expected: True

    # Find representatives
    print(f"Representative of 0: {uf.find(0)}")
    print(f"Representative of 1: {uf.find(1)}")
    print(f"Representative of 2: {uf.find(2)}")
    print(f"Representative of 3: {uf.find(3)}")
    print(f"Representative of 4: {uf.find(4)}")
    print(f"Representative of 5: {uf.find(5)}")

    # Count the number of sets
    print(f"Number of distinct sets: {uf.count_sets()}")  # Expected: 1


    # Example using the basic UnionFind class
    print("\n--- Basic Union-Find Example ---")
    uf_basic = UnionFind(5)
    uf_basic.union(0, 1)
    uf_basic.union(3, 4)
    print(f"Are 0 and 1 connected? {uf_basic.isConnected(0, 1)}") # True
    print(f"Are 0 and 2 connected? {uf_basic.isConnected(0, 2)}") # False
    print(f"Number of sets: {uf_basic.count_sets()}")

    # Example using UnionFindRanked
    print("\n--- Union-Find with Rank Example ---")
    uf_ranked = UnionFindRanked(5)
    uf_ranked.union(0, 1)
    uf_ranked.union(2, 3)
    uf_ranked.union(1, 3)
    print(f"Are 0 and 3 connected? {uf_ranked.isConnected(0, 3)}") # True
    print(f"Number of sets: {uf_ranked.count_sets()}")

    # Example using UnionFindIterativePC
    print("\n--- Union-Find with Iterative Path Compression ---")
    uf_iterative = UnionFindIterativePC(5)
    uf_iterative.union(0, 1)
    uf_iterative.union(2, 3)
    uf_iterative.union(1, 3)
    print(f"Are 0 and 3 connected? {uf_iterative.isConnected(0, 3)}")
    print(f"Number of sets: {uf_iterative.count_sets()}")

if __name__ == "__main__":
    main()

File: JavaScript/8-Graphs(BFS-DFS)/Algorithms/7-Union-Find(DisjointSetUnion-DSU)/Union-Find(DisjointSetUnion-DSU).js
/**
 * Union-Find (Disjoint Set Union - DSU) Implementation in JavaScript
 * Demonstrates 5 different approaches:
 * 1. Quick Find
 * 2. Quick Union
 * 3. Weighted Quick Union
 * 4. Path Compression (with Weighted Quick Union)
 * 5. Rank with Path Compression
 */

// Helper function to create an array of size n with initial values
function createArray(n, initialValue) {
  const arr = new Array(n);
  for (let i = 0; i < n; i++) {
    arr[i] = initialValue;
  }
  return arr;
}

// ----------------------------------------------------------------------
// 1. Quick Find
// ----------------------------------------------------------------------
class QuickFind {
  constructor(n) {
    this.id = createArray(n, (i) => i); // Initialize each element as its own root
    this.count = n; // Number of connected components
  }

  find(p) {
    return this.id[p]; // O(1)
  }

  union(p, q) {
    const pId = this.find(p);
    const qId = this.find(q);

    if (pId === qId) return; // Already connected

    for (let i = 0; i < this.id.length; i++) {
      if (this.id[i] === pId) {
        this.id[i] = qId; // Change all elements with pId to qId
      }
    }
    this.count--;
  }

  isConnected(p, q) {
    return this.find(p) === this.find(q);
  }

  countSets() {
    return this.count;
  }
}

// ----------------------------------------------------------------------
// 2. Quick Union
// ----------------------------------------------------------------------
class QuickUnion {
  constructor(n) {
    this.parent = createArray(n, (i) => i); // Initialize each element as its own parent
    this.count = n;
  }

  find(p) {
    // Find the root of p
    while (p !== this.parent[p]) {
      p = this.parent[p];
    }
    return p;
  }

  union(p, q) {
    const pRoot = this.find(p);
    const qRoot = this.find(q);

    if (pRoot === qRoot) return; // Already connected

    this.parent[pRoot] = qRoot; // Make the root of p point to the root of q
    this.count--;
  }

  isConnected(p, q) {
    return this.find(p) === this.find(q);
  }

  countSets() {
    return this.count;
  }
}

// ----------------------------------------------------------------------
// 3. Weighted Quick Union
// ----------------------------------------------------------------------
class WeightedQuickUnion {
  constructor(n) {
    this.parent = createArray(n, (i) => i); // Initialize each element as its own parent
    this.size = createArray(n, 1); // Size of each tree, initially 1
    this.count = n;
  }

  find(p) {
    // Find the root of p
    while (p !== this.parent[p]) {
      p = this.parent[p];
    }
    return p;
  }

  union(p, q) {
    const pRoot = this.find(p);
    const qRoot = this.find(q);

    if (pRoot === qRoot) return; // Already connected

    // Make the smaller tree point to the larger tree
    if (this.size[pRoot] < this.size[qRoot]) {
      this.parent[pRoot] = qRoot;
      this.size[qRoot] += this.size[pRoot];
    } else {
      this.parent[qRoot] = pRoot;
      this.size[pRoot] += this.size[qRoot];
    }
    this.count--;
  }

  isConnected(p, q) {
    return this.find(p) === this.find(q);
  }
  countSets() {
    return this.count;
  }
}

// ----------------------------------------------------------------------
// 4. Path Compression (with Weighted Quick Union)
// ----------------------------------------------------------------------
class PathCompression {
  constructor(n) {
    this.parent = createArray(n, (i) => i);
    this.size = createArray(n, 1);
    this.count = n;
  }

  find(p) {
    // Find the root of p with path compression
    while (p !== this.parent[p]) {
      this.parent[p] = this.parent[this.parent[p]]; // Path compression: make p point to its grandparent
      p = this.parent[p];
    }
    return p;
  }

  union(p, q) {
    const pRoot = this.find(p);
    const qRoot = this.find(q);

    if (pRoot === qRoot) return;

    if (this.size[pRoot] < this.size[qRoot]) {
      this.parent[pRoot] = qRoot;
      this.size[qRoot] += this.size[pRoot];
    } else {
      this.parent[qRoot] = pRoot;
      this.size[pRoot] += this.size[qRoot];
    }
    this.count--;
  }

  isConnected(p, q) {
    return this.find(p) === this.find(q);
  }
  countSets() {
    return this.count;
  }
}

// ----------------------------------------------------------------------
// 5. Rank with Path Compression
// ----------------------------------------------------------------------
class RankPathCompression {
  constructor(n) {
    this.parent = createArray(n, (i) => i);
    this.rank = createArray(n, 0); // Use rank instead of size
    this.count = n;
  }

  find(p) {
    while (p !== this.parent[p]) {
      this.parent[p] = this.parent[this.parent[p]];
      p = this.parent[p];
    }
    return p;
  }

  union(p, q) {
    const pRoot = this.find(p);
    const qRoot = this.find(q);

    if (pRoot === qRoot) return;

    if (this.rank[pRoot] < this.rank[qRoot]) {
      this.parent[pRoot] = qRoot;
    } else if (this.rank[pRoot] > this.rank[qRoot]) {
      this.parent[qRoot] = pRoot;
    } else {
      this.parent[qRoot] = pRoot; // Make qRoot child of pRoot
      this.rank[pRoot]++; // Increase rank of pRoot
    }
    this.count--;
  }

  isConnected(p, q) {
    return this.find(p) === this.find(q);
  }

  countSets() {
    return this.count;
  }
}

// ----------------------------------------------------------------------
// Example Usage and Testing
// ----------------------------------------------------------------------

function runTests() {
  const n = 10; // Number of elements

  console.log("--------------------------------------------------");
  console.log("1. Quick Find");
  console.log("--------------------------------------------------");
  const qf = new QuickFind(n);
  console.log("Initial:", qf.id);
  qf.union(0, 1);
  qf.union(2, 3);
  qf.union(4, 5);
  qf.union(6, 7);
  qf.union(8, 9);
  console.log("After unions (0,1), (2,3), (4,5), (6,7), (8,9):", qf.id);
  console.log("isConnected(0, 1):", qf.isConnected(0, 1)); // true
  console.log("isConnected(0, 2):", qf.isConnected(0, 2)); // false
  console.log("Number of sets:", qf.countSets()); //5

  console.log("\n--------------------------------------------------");
  console.log("2. Quick Union");
  console.log("--------------------------------------------------");
  const qu = new QuickUnion(n);
  console.log("Initial parent:", qu.parent);
  qu.union(0, 1);
  qu.union(2, 3);
  qu.union(4, 5);
  qu.union(6, 7);
  qu.union(8, 9);
  console.log("After unions (0,1), (2,3), (4,5), (6,7), (8,9):", qu.parent);
  console.log("isConnected(0, 1):", qu.isConnected(0, 1));
  console.log("isConnected(0, 2):", qu.isConnected(0, 2));
  console.log("Number of sets:", qu.countSets());

  console.log("\n--------------------------------------------------");
  console.log("3. Weighted Quick Union");
  console.log("--------------------------------------------------");
  const wqu = new WeightedQuickUnion(n);
  console.log("Initial parent:", wqu.parent);
  console.log("Initial sizes:", wqu.size);
  wqu.union(0, 1);
  wqu.union(2, 3);
  wqu.union(4, 5);
  wqu.union(6, 7);
  wqu.union(8, 9);
  wqu.union(0, 2); // Connect 0 and 2
  console.log("After unions (0,1), (2,3), (4,5), (6,7), (8,9), (0,2):", wqu.parent);
  console.log("Sizes:", wqu.size);
  console.log("isConnected(0, 2):", wqu.isConnected(0, 2));
  console.log("Number of sets:", wqu.countSets()); // 4

  console.log("\n--------------------------------------------------");
  console.log("4. Path Compression");
  console.log("--------------------------------------------------");
  const pc = new PathCompression(n);
  pc.union(0, 1);
  pc.union(2, 3);
  pc.union(4, 5);
  pc.union(6, 7);
  pc.union(8, 9);
  pc.union(0, 2);
  console.log("Initial parent:", pc.parent);
  pc.find(1); // Perform find to trigger path compression.
  console.log("After unions and find(1):", pc.parent);
  console.log("isConnected(0, 2):", pc.isConnected(0, 2));
  console.log("Number of sets:", pc.countSets());

  console.log("\n--------------------------------------------------");
  console.log("5. Rank with Path Compression");
  console.log("--------------------------------------------------");
  const rpc = new RankPathCompression(n);
  rpc.union(0, 1);
  rpc.union(2, 3);
  rpc.union(4, 5);
  rpc.union(6, 7);
  rpc.union(8, 9);
  rpc.union(0, 2);
  console.log("Initial parent:", rpc.parent);
  console.log("Initial ranks:", rpc.rank);
  rpc.find(1);
  console.log("After unions and find(1):", rpc.parent);
  console.log("Ranks:", rpc.rank);
  console.log("isConnected(0, 2):", rpc.isConnected(0, 2));
  console.log("Number of sets:", rpc.countSets());
}

runTests();
